### PR TITLE
executor: don't reuse Executor in IndexLookUpJoin, remove doRequestForDatums()

### PR DESCRIPTION
**Pull request by tiancaiamao**
I change the `DataReader` interface, let it return a new Executor object every time:

```
type DataReader interface {
 doRequestForDatums(goCtx goctx.Context, datums [][]types.Datum) (Executor, error)
}
```

The returned Executor is not reused...

Remove `doRequestForDatums()` method from `TableReaderExecutor` `IndexReaderExecutor` `IndexLookUpExecutor` , and implement `DataReader` by introducing new struct and wrap around them....

@coocood @zz-jason

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -1093,3 +1105,93 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
 	}
 	return e
 }
+
+// DataReaderBuilder build a executor.
+// The executor can be used to read data in the ranges which are constructed by datums.
+type DataReaderBuilder interface {
```
Comment (zz-jason) [inline executor/builder.go]: no need to export `DataReaderBuilder`

File: executor/builder.go [hunk 2]
```diff
@@ -1093,3 +1093,83 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+		e := builder.executorBuilder.buildTableReader(v)
+		return buildTableReaderForDatums(goCtx, e, datums)
+	case *plan.PhysicalIndexLookUpReader:
+		e := builder.executorBuilder.buildIndexLookUpReader(v)
+		return buildIndexLookUpReaderForDatums(goCtx, e, datums)
+	}
+	return nil, errors.New("should never run here")
+}
+
+func buildTableReaderForDatums(goCtx goctx.Context, e *TableReaderExecutor, datums [][]types.Datum) (Executor, error) {
```
Comment (coocood) [inline executor/builder.go]: This should be `dataReaderBuilder`'s method.

File: executor/builder.go [hunk 3]
```diff
@@ -1093,3 +1093,83 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+	handles := make([]int64, 0, len(datums))
+	for _, datum := range datums {
+		handles = append(handles, datum[0].GetInt64())
+	}
+	return doRequestForHandles(goCtx, e, handles)
+}
+
+func doRequestForHandles(goCtx goctx.Context, e *TableReaderExecutor, handles []int64) (Executor, error) {
```
Comment (coocood) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 4]
```diff
@@ -1093,3 +1093,83 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+	}
+	e.result, err = distsql.SelectDAG(goCtx, e.ctx.GetClient(), kvReq, e.schema.Len())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	e.result.Fetch(goCtx)
+	return e, nil
+}
+
+func buildIndexReaderForDatums(goCtx goctx.Context, e *IndexReaderExecutor, values [][]types.Datum) (Executor, error) {
```
Comment (coocood) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 5]
```diff
@@ -1093,3 +1093,83 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+	}
+	e.result, err = distsql.SelectDAG(e.ctx.GoCtx(), e.ctx.GetClient(), kvReq, e.schema.Len())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	e.result.Fetch(goCtx)
+	return e, nil
+}
+
+func buildIndexLookUpReaderForDatums(goCtx goctx.Context, e *IndexLookUpExecutor, values [][]types.Datum) (Executor, error) {
```
Comment (coocood) [inline executor/builder.go]: ditto

**Comment by tiancaiamao** [comment]
PTAL @coocood @zz-jason

File: executor/builder.go [hunk 6]
```diff
@@ -1094,6 +1092,90 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+func (builder *dataReaderBuilder) buildTableReaderForDatums(goCtx goctx.Context, v *plan.PhysicalTableReader, datums [][]types.Datum) (Executor, error) {
+	e := builder.executorBuilder.buildTableReader(v)
+	handles := make([]int64, 0, len(datums))
+	for _, datum := range datums {
+		handles = append(handles, datum[0].GetInt64())
+	}
+	return builder.doRequestForHandles(goCtx, e, handles)
+}
+
+func (builder *dataReaderBuilder) doRequestForHandles(goCtx goctx.Context, e *TableReaderExecutor, handles []int64) (Executor, error) {
```
Comment (coocood) [inline executor/builder.go]: `buildTableReaderForHandles`

File: executor/builder.go [hunk 7]
```diff
@@ -1094,6 +1092,90 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+	sort.Sort(int64Slice(handles))
```
Comment (zz-jason) [inline executor/builder.go]: use `Int64Slice` in package **github.com/cznic/sortutil**

File: executor/builder.go [hunk 8]
```diff
@@ -946,21 +954,11 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plan.PhysicalIndexJoin) Execut
 
 	// for IndexLookUpJoin, left is always the outer side.
 	outerExec := b.build(v.Children()[0])
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/builder.go]: It's done in executor/adaptor.go `buildExecutor()`

File: executor/builder.go [hunk 9]
```diff
@@ -1094,6 +1092,90 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+	e.result, err = distsql.SelectDAG(goCtx, e.ctx.GetClient(), kvReq, e.schema.Len())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	e.result.Fetch(goCtx)
+	return e, nil
+}
+
+func (builder *dataReaderBuilder) buildIndexReaderForDatums(goCtx goctx.Context, v *plan.PhysicalIndexReader, values [][]types.Datum) (Executor, error) {
+	e := builder.executorBuilder.buildIndexReader(v)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/builder.go]: done

File: executor/builder.go [hunk 10]
```diff
@@ -1094,6 +1092,90 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
+	e.result, err = distsql.SelectDAG(e.ctx.GoCtx(), e.ctx.GetClient(), kvReq, e.schema.Len())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	e.result.Fetch(goCtx)
+	return e, nil
+}
+
+func (builder *dataReaderBuilder) buildIndexLookUpReaderForDatums(goCtx goctx.Context, v *plan.PhysicalIndexLookUpReader, values [][]types.Datum) (Executor, error) {
+	e := builder.executorBuilder.buildIndexLookUpReader(v)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/builder.go]: done

File: executor/distsql.go [hunk 11]
```diff
@@ -868,21 +796,23 @@ func (e *IndexLookUpExecutor) executeTask(task *lookupTableTask, goCtx goctx.Con
...
+	e1, err = e.builder.doRequestForHandles(goCtx, &TableReaderExecutor{
 		table:   e.table,
 		tableID: e.tableID,
 		dagPB:   e.tableRequest,
 		schema:  schema,
 		ctx:     e.ctx,
-	}
-	err = tableReader.doRequestForHandles(goCtx, task.handles)
+	}, task.handles)
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/distsql.go]: done

**Comment by coocood** [comment]
LGTM

**Pull request merged**